### PR TITLE
Add XBOX authentication

### LINKED ISSUE #1462: MCPE Crashed
The PocketMine Modified No Supports maps installed
You Join In The Map And Crashe Your MCPE
 nobody can put maps of mini-games or Spawn
 the pc and android
PHP7
0.15.X
Android
PC
...

### LINKED ISSUE #1642: Feature Request: Xbox function
#### Issue description

Could you please add a function for checking whether or not a player is using xbox? Look here: https://github.com/ClearSkyTeam/ClearSky/commit/b2baec8510e35009c238f4e1bf04b0cd428bfffb
#### OS and versions

* Genisys:
* PHP:
* Server OS:
* Game version: 

#### Crashdump, backtrace, memory dumps, plugins list or other files

```
```

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
Close #1642 
New dependency: OpenSSL is now mandantory.

### CODE REVIEW

**Comment by XFuryMCPE** [comment]
Omg, goodbye login plugins!

**Comment by dktapps** [comment]
@XFuryMCPE don't forget banned players changing names :package: 
I've yet to test this, but it looks good.

**Comment by ishitatsuyuki** [comment]
Additional information:
The UUID of the players are constant.
This works on Docker installation, but needs additional configuration on Windows.

**Comment by Muqsit** [comment]
Any way to add support for gamertags with spaces?

**Comment by dktapps** [comment]
Hmm. I can't login, with or without Xbox connected. I get disconnection screen every time, whether I'm logged in or not.

**Comment by ishitatsuyuki** [comment]
Please try again. By the way, shoghi is adding invalid thingy in the last of chain, so I added a hack and some comments.

**Comment by FaanMaario** [comment]
Seems to work, I can tchat and play on my server.

**Comment by MrDoni98** [comment]
`Error: Call to undefined function pocketmine\utils\openssl_random_pseudo_bytes()`
`File: /src/pocketmine/utils/Utils`
`Line: 438`
`Type: notice`
`Code:`
`[429] 					$systemRandom = fread($fp, 64);`
`[430] 					fclose($fp);`
`[431] 				}else{`
`[432] 					$systemRandom = str_repeat("\x00", 64);`
`[433] 				}`
`[434] `
`[435] 				$strongEntropyValues = [`
`[436] 					is_array($startEntropy) ? hash("sha512", $startEntropy[($rounds +` `$drop)%count($startEntropy)], true) : hash("sha512", $startEntropy, true), //Get a random index of the` `startEntropy, or just read it`
`[437] 					$systemRandom,`
`[438] 					openssl_random_pseudo_bytes(64),`
`[439] 					$value,`
`[440] 				]`

**Comment by ishitatsuyuki** [comment]
@MrDoni98 You have no OpenSSL extension installed. Read the PR description.

**Comment by MrDoni98** [comment]
@ishitatsuyuki ok thanks

**Comment by MrDoni98** [comment]
@ishitatsuyuki  how install this? 
"php -i |grep -A 5 openssl " says that OpenSSL enabled

**Comment by ishitatsuyuki** [comment]
@MrDoni98 please move to chat for support questions.

**Comment by dktapps** [comment]
Lol, always #blameshoghi. I'll retest.

**Comment by dktapps** [comment]
It now seems functional, but I'm encountering crashes with plugins listening for PlayerLoginEvent.
```
[10:10:20] [Server thread/CRITICAL]: "Could not pass event 'pocketmine\event\player\PlayerLoginEvent' to 'PurePerms v1.3.7-patch1': Call to a member function clearPermissions() on boolean on _64FF00\PurePerms\PPListener
[10:10:20] [Server thread/CRITICAL]: Error: "Call to a member function clearPermissions() on boolean" (EXCEPTION) in "/plugins/PurePerms_v1.3.7-patch1.phar/src/_64FF00/PurePerms/PurePerms" at line 724
```

I believe that this event should not be called if the player is not authenticated.

**Comment by dktapps** [comment]
Also on server shutdown:
```
[10:18:11] [Server thread/CRITICAL]: TypeError: "Argument 1 passed to pocketmine\Player::removeAttachment() must be an instance of pocketmine\permission\PermissionAttachment, boolean given, called in phar://C:/Users/Admin/Downloads/Genisys/plugins/PurePerms_v1.3.7-patch1.phar/src/_64FF00/PurePerms/PurePerms.php on line 755" (EXCEPTION) in "/src/pocketmine/Player" at line 657
```

File: src/pocketmine/Player.php [hunk 1]
```diff
@@ -2368,6 +2368,10 @@ public function handleDataPacket(DataPacket $packet){
 				$this->iusername = strtolower($this->username);
 				$this->protocol = $packet->protocol;
 
+				if($this->server->getConfigBoolean("online-mode", false) && $packet->identityPublicKey === null){
+					$this->kick("disconnectionScreen.notAuthenticated", false);
```
Comment (dktapps) [inline src/pocketmine/Player.php]: I think this needs to be close() and not kick(). See below disconnection screens.

Comment (ishitatsuyuki) [inline src/pocketmine/Player.php]: Agree with break. Closing the session won't show the reason to user, thus confusing.

Comment (dktapps) [inline src/pocketmine/Player.php]: Done

**Comment by dktapps** [comment]
Seems stable now. +1.

Off-topic @64FF00: This message from PurePerms gave me a laugh: `[10:26:52] [Server thread/DEBUG]: [PurePerms] Why did you give me an invalid unique id? *cries* (userName: ~~[redacted]~~, isConnected: false, isOnline: false, isValid: true)` :laughing:

**Comment by ishitatsuyuki** [comment]
Checklist before merge:
[ ] * server.properties default added
[ ] * New PHP for Windows uploaded
[ ] * New PHP for ARM uploaded

**Comment by MrDoni98** [comment]
I did not find where to get the OpenSSL extension, tell me please

**Comment by ishitatsuyuki** [comment]
@MrDoni98 Please come to one of the chat so we can communicate conveniently.

**Comment by MrDoni98** [comment]
@ishitatsuyuki what chat?

**Comment by MrDoni98** [comment]
gitter?

**Comment by ishitatsuyuki** [comment]
@MrDoni98 join and ask, not here.

File: src/pocketmine/Player.php (hunk 1 above)
Comment (robske110) [inline src/pocketmine/Player.php]: And close will show the reason too. Kick is calling close

**Comment by robske110** [comment]
Just to ask it open: may i copy (cherry-pick) your commits/changes. Of course with proper acknowledgement.

**Comment by ishitatsuyuki** [comment]
@robske110 Following GPL, I don't restrict anyone from copying.

**Comment by dktapps** [comment]
Okay, I think this is good to go. +1.

[Approved with PullApprove](https://pullapprove.com/iTXTech/Genisys/pull-request/1655/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=iTXTech/Genisys)

**Comment by ishitatsuyuki** [comment]
@iTXTech/trusted Come on, let's take action.

**Comment by CrayonDing** [comment]
+1.Good job.

**Pull request merged**